### ISSUE TITLE
Can Spring MVC adapters be provided?

### ISSUE THREAD

**Issue by kaizi2009** [alibaba/Sentinel#1074]
### Can Spring MVC adapters be provided? Used to support path variable.
```java
@GetMapping("/foo/{id}")
public String apiFoo(@PathVariable("id") Long id) {
    return "Hello " + id;
}
```
#### I want to do this .

**Comment by sczyh30** [comment]
Hi, you may refer to https://github.com/alibaba/Sentinel/issues/286#issuecomment-444848063

Contributions are welcomed!

**Comment by zhaoyuguang** [comment]
I finish it and assign it to me.

**Comment by sczyh30** [comment]
> I finish it and assign it to me.

Nice, looking forward to your PR!

**Comment by sczyh30** [comment]
Any progress? :)

**Comment by zhaoyuguang** [comment]
> Any progress? :)

遇到了一些问题，由于资源是基于树，且唯一，并且拦截器是晚于过滤器，我的代码是这样的，
```
public class BestPatternInterceptor extends HandlerInterceptorAdapter {

    @Override
    public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler) throws Exception {
        String target = (String) request.getAttribute(HandlerMapping.BEST_MATCHING_PATTERN_ATTRIBUTE);
        if (StringUtil.isEmpty(target)) {
            return true;
        }
        Sout.sout("handle url start", target);
        Entry entry = null;
        try {
            UrlCleaner urlCleaner = WebCallbackManager.getUrlCleaner();
            if (urlCleaner != null) {
                target = urlCleaner.clean(target);
            }
            if (!StringUtil.isEmpty(target)) {
                entry = SphU.entry(target, EntryType.IN);
            }
            return true;
        } catch (BlockException ex) {
            WebCallbackManager.getUrlBlockHandler().blocked(request, response, ex);
            return false;
        } finally {
            if (entry != null) {
                entry.exit();
            }
        }
    }
}
```
会造成`/foo/{id}`,关联到`/foo/1`，这个节点，如果`/foo/2`请求出现，就不会在`/foo/2`下出现`/foo/{id}`

**Comment by sczyh30** [comment]
First, it's not recommended to use both the Servlet `CommonFilter` and SentinelHandlerInterceptor adapter at the same time. You could also regard this HandlerInterceptor as the entrance (i.e. call `ContextUtil.enter("sentinel_spring_web_context")` here).

Second, you have to exit the entry in `afterCompletion` (or `postHandle`), or the response time will not be correct. You may carry the `Entry` in the request attribute like this: `request.setAttribute(SENTINEL_ENTRY_ATTR_KEY, entry)`. You also have to pay attention to the thread context of `preHandle` and `afterCompletion` for the same `Entry`.

A simple code snippet:

```java
@Override
public void afterCompletion(HttpServletRequest request, HttpServletResponse response, Object handler, Exception ex) throws Exception {
    Object e = request.getAttribute(SENTINEL_ENTRY_ATTR_KEY);
    if (e == null) {
        return;
    }
    Entry entry = (Entry) e;
    if (ex != null) {
        Tracer.traceEntry(ex, entry);
    }
    entry.exit();
    request.removeAttribute(SENTINEL_ENTRY_ATTR_KEY);
}
```

**Comment by zhaoyuguang** [comment]
@sczyh30  Thanks lot~
Just need to write clearly in the document, is it not recommended to use the Servlet `CommonFilter` and `SentinelHandlerInterceptor` at the same time?
Still doing this in the `sentinel-web-servlet` module? Still writing an adaptor, just like `sentinel-spring-webflux-adapter`？

**Comment by sczyh30** [comment]
Resolved in #1104
